cmake/KokkosConfigCommon.cmake.in

```diff
@@ -203,6 +204,573 @@ int main()
+            )
+            if(depends_linking)
+              set(link_dependency ON)
+            endif()
+
+            #if a dependency is infectious
+            if(depends_transitively)
+              set(compile_dependency ON)
+              #and if we link PUBLIC to it we get infectious, too
+              if(target_interface_dependencies)
+                if(${dependency} IN_LIST target_interface_dependencies)
+                  set(transitive_dependency ON)
+                endif()
+              endif()
+            endif()
+          endforeach()
+        endif()
+      endif()
+    endif()
+  endif()
+
+  set(${COMPILE_DEPENDENCY} ${compile_dependency} PARENT_SCOPE)
+  set(${TRANSITIVE_COMPILE_DEPENDENCY} ${transitive_dependency} PARENT_SCOPE)
+  set(${LINK_DEPENDENCY} ${link_dependency} PARENT_SCOPE)
+
+  if(link_dependency)
+    list(APPEND ${COMPILE_DEPENDENCY_CACHE} ${INVESTIGATED_TARGET})
+  endif()
+  set(${COMPILE_DEPENDENCY_CACHE} "${${COMPILE_DEPENDENCY_CACHE}}" PARENT_SCOPE)
+
+  if(transitive_dependency)
+    list(APPEND ${TRANSITIVE_COMPILE_DEPENDENCY_CACHE} ${INVESTIGATED_TARGET})
+  endif()
+  set(${TRANSITIVE_COMPILE_DEPENDENCY_CACHE} "${${TRANSITIVE_COMPILE_DEPENDENCY_CACHE}}" PARENT_SCOPE)
+endfunction()
+
+# this function searches a given directory recursively finding all targets that link to DEPENDENT_LIBRARY
+#
+#       FOUND_TARGETS         --> output: list of all found targets that link to kokkos and need our compiler
+#       DEPENDENT_LIBRARY     --> input: The library to search for in the target properties
+#       SEARCH_DIR            --> input: the root dir to start the recursion from
+#       EXCLUDE_DIRS          --> input: list of directories that should be excluded
+#
+function(kokkos_get_directory_targets)
+  cmake_parse_arguments(INP "" "SEARCH_DIR" "FOUND_TARGETS;EXCLUDE_DIRS" ${ARGN})
+  set(targets "")
+  set(subdirs "")
+
+  if(${INP_SEARCH_DIR} IN_LIST INP_EXCLUDE_DIRS)
+    message(
+      DEBUG
+      "Kokkos excludes dir ${INP_SEARCH_DIR} from setting build options as it was marked to be excluded by `kokkos_exclude_from_setting_build_properties`"
+    )
+  else()
+    get_property(targets DIRECTORY ${INP_SEARCH_DIR} PROPERTY BUILDSYSTEM_TARGETS)
+    get_property(subdirs DIRECTORY ${INP_SEARCH_DIR} PROPERTY SUBDIRECTORIES)
+  endif()
+
+  # recursively visit subdirs
+  if(subdirs)
+    foreach(subdir ${subdirs})
+      kokkos_get_directory_targets(FOUND_TARGETS subdir_targets SEARCH_DIR ${subdir} EXCLUDE_DIRS ${INP_EXCLUDE_DIRS})
+      list(APPEND targets ${subdir_targets})
+    endforeach()
+  endif()
+
+  # push results out of recursive function
+  set(${INP_FOUND_TARGETS} ${targets} PARENT_SCOPE)
+endfunction()
+
+# this function searches a given directory recursively finding all targets that link to DEPENDENT_LIBRARY
+#
+#       GLOBAL         --> input: exclude everything from Kokkos setting build properties
+#       DIRECTORY      --> input: list of directories that should be excluded
```

Comment:
It is not clear if we can pass a list (so `DIRECTORIES`) and if we can call this function several times.

Reply:
I chose the naming similar to https://cmake.org/cmake/help/latest/command/set_property.html. Here `TARGET`, `SOURCE`, etc. take multiple arguments. But I see your argument and am fine with either singular or plural.